contract/src/RadicleCloudBase.sol

```diff
@@ -0,0 +1,269 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity ^0.8.6;
+
+import "./ImpactPoints.sol";
+
+/// @title RadicleCloudBase does on-chain accounting and emits events for operator.
+/// @dev RadicleCloudBase is an abstract contract.
+abstract contract RadicleCloudBase {
+    /// @notice Owner is needed for changing rate and withdrawing funds.
+    address public owner;
+
+    /// @dev Keep info for tracking each deployment
+    struct Deployment {
+        uint64 start;
+        uint64 expiry;
+        uint64 rate;
+        address owner;
+    }
+    /// @notice Info regading each deployment (org)
+    mapping(address => Deployment) public dep;
+
+    /// @notice Rate (price) new buyers are charged
+    uint64 public ratePerBlock;
+    /// @notice Rate (price) currently being earned
+    int64 private runningRate = 0;
+
+    /// @notice Duration of package, roughly equivalent to one month
+    uint32 public immutable duration; // 185_142
+    /// @notice Blocks to wait before being eligible for withdraw
+    uint32 public immutable withdrawWait; // 6170
+
+    /// @notice MAXBLOCK is the maximum value of uint64
+    uint64 private constant MAXBLOCK = type(uint64).max;
+
+    /// @dev Add functionalities for index and block retrieval
+    using ImpactPoints for uint64[];
+    /// @notice Special array of blocks when buys happened
+    uint64[] private buys = [1, MAXBLOCK];
+    /// @notice Special array of blocks when expires happen
+    uint64[] private expires = [1, MAXBLOCK];
+    /// @notice Special array of blocks when cancels happened
+    uint64[] private cancels = [1, MAXBLOCK];
+    /// @notice Special array of blocks when renews happened
+    uint64[] private renews = [1, MAXBLOCK];
+    /// @notice Whether a renewal happened at this block
+    mapping(uint64 => bool) private hasRenewal;
```

Comment:
The assumption here is that we don't want to introduce the `DripsHub`'s `cycles`. Then realizing revenue would have cost independent from the number of `topUp`s, BUT it would obviously make the realizable amount updated only once per cycle, which may or may not be an acceptable trade-off.

Reply:
While simpler, I'm not convinced this would be cheaper in terms of gas. We should expect 1:2 ratio of long-term to short-term buyers, and probably a closer ratio due to gas, and DAOs. This means we'll end up in second branch a good portion of time, as operator would probably like to withdraw on a monthly basis and some users would be on a yearly package. Whereas with a running rate system, there's no writes to storage, a bunch of `min`s and a multiplication most of the time.

Reply:
Let's break down what happens if 10 `org`s top up for 12 months and then somebody realizes once every month.

In the **current** implementation each `org` in `topUp` will write 12 elements to `hasRenewal`. We can't assume that they'll be in sync, so there won't be overlaps in which `hasRenewal` keys will be used and usually all writes will be first-time. Additionally there are 2 `impact` entries written, one may be overwritten and 1 guaranteed to be first-time.

In the **proposed** implementation each `org` in `topUp` updates 1 `chunksLen` slot and uses 1 fresh slot in `chunks`.

The proposed implementation is clearly cheaper in `topUp`, by ~250K gas for each 12-month top-up.

When it comes to `_calcRealizedRevenue`, I agree that in a perfect run the **current** implementation is going to be slightly cheaper. It's going to cold-read 9 slots (`runningRate` and 2 reads for each of `buys`, `cancels`, `renews` and `expires`),  then update 1 slot (`lastProcessed`). That's ~ 24K gas. It'll get much more expensive if there have been any renewals.

On the other hand the **proposed** version writes 11 slots (1 length and 10 chunks), which is ~55K gas. It scales linearly with the number of orgs.

In this particular case the **proposed** version is ~30K bas more expensive for each revenue realization.

The question is how many orgs are we expecting to see and how long the support periods are going to be?